### PR TITLE
 WL-1881 | Added a management command to sync the ecommerce data to hubspot.

### PR DESCRIPTION
**Pull request by hasnain-naveed**

### CODE REVIEW

**Comment by hasnain-naveed** [comment]
@HammadAhmadWaqas this is final PR.

File: ecommerce/core/management/commands/sync_hubspot.py [hunk 1]
```diff
@@ -0,0 +1,439 @@
...
+class Command(BaseCommand):
+    help = 'Sync Product, Orders and Lines to partners Hubspot server.'
+
+    def _get_hubspot_enable_sites(self):
+        """
+        Returns all SiteConfigurations which have hubspot enabled
+        """
+        return SiteConfiguration.objects.filter(hubspot_secret_key__isnull=False)
+
+    def _hubspot_endpoint(self, api, api_url, api_type, body=None, **kwargs):
```
Comment (ziafazal) [inline ecommerce/core/management/commands/sync_hubspot.py]: Can we make signature of this method a little more readable.
i.e `def _hubspot_endpoint(self, hubspot_object, api_url, method, body=None, **kwargs):`

File: ecommerce/core/management/commands/sync_hubspot.py [hunk 2]
```diff
@@ -0,0 +1,439 @@
...
+            if orders:
+                self.stdout.write(
+                    'Pulled unsynced orders for site {site} from {last_sync}'.format(
+                        site=site_configuration.site.domain, last_sync=last_synced_order.date_placed
+                    )
+                )
+        else:
+            orders = orders.filter(date_placed__lte=datetime.now())
+            self.stdout.write(
+                'It is first syncing call, it will pull all the orders for syncing for site {site}'.format(
```
Comment (ziafazal) [inline ecommerce/core/management/commands/sync_hubspot.py]: May be we can change the wording `No last synced order found...`

**Review by ziafazal — APPROVED** [review]
@hasnain-naveed overall looks good just few nits.

File: ecommerce/core/management/commands/sync_hubspot.py [hunk 3]
```diff
@@ -0,0 +1,455 @@
...
+            self._upsert_hubspot_objects(
+                LINE_ITEM,
+                self._get_hubspot_line_item_structure(unsynced_order_lines),
+                site_configuration
+            )
+        else:
+            self.stdout.write('No data found to sync for site {site}'.format(site=site_configuration.site.domain))
+
+    def add_arguments(self, parser):
+
```
Comment (georgebabey) [inline ecommerce/core/management/commands/sync_hubspot.py]: nit: space

File: ecommerce/core/management/commands/sync_hubspot.py [hunk 4]
```diff
@@ -0,0 +1,455 @@
...
+    ORDER.COMPLETE: "processed"
+}
+PRODUCT = "PRODUCT"
+LINE_ITEM = "LINE_ITEM"
+DEAL = "DEAL"
+BATCH_SIZE = 200
+
+
+class Command(BaseCommand):
+    help = 'Sync Product, Orders and Lines to partners Hubspot server.'
```
Comment (georgebabey) [inline ecommerce/core/management/commands/sync_hubspot.py]: I would take `partner` out of here as that is an edx specific term. Maybe `Sync Products, Orders and Line Items to a Hubspot server.`

File: ecommerce/core/migrations/0048_siteconfiguration_hubspot_secret_key.py [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+
+    dependencies = [
+        ('core', '0047_businessclient_enterprise_customer_uuid'),
+    ]
+
+    operations = [
+        migrations.AddField(
+            model_name='siteconfiguration',
+            name='hubspot_secret_key',
+            field=models.CharField(blank=True, help_text='Secret key for Hubspot portal authentication', max_length=255, null=True, verbose_name='Hubspot Portal Secret Key'),
```
Comment (georgebabey) [inline ecommerce/core/migrations/0048_siteconfiguration_hubspot_secret_key.py]: General django best practice is to not allow null for CharField (see https://docs.djangoproject.com/en/dev/ref/models/fields/#null) 

Right now you are checking for non-`null` to determine if hubspot is enabled, but new configurations will have this set to empty string.  https://github.com/edx/ecommerce/pull/2146/files#diff-344c0546bfd4c3a65a58efc931f82f3aR147

I'd recommend changing this to `null=False`

**Review by georgebabey — CHANGES_REQUESTED** [review]
Looks great overall but I believe there is one issue with null vs empty_string on the configuration

File: ecommerce/core/migrations/0048_siteconfiguration_hubspot_secret_key.py (hunk 5 above)
Comment (hasnain-naveed) [reply ecommerce/core/migrations/0048_siteconfiguration_hubspot_secret_key.py]: @georgebabey WL-1880 is pre-requisite of WL-1881. WL-1880 covers this model change. I have updated my [PR](https://github.com/edx/ecommerce/pull/2118) according to your suggestion. Waiting for your +1. Thanks.

Comment (georgebabey) [reply ecommerce/core/migrations/0048_siteconfiguration_hubspot_secret_key.py]: Merged WL-1880 in, so this will need to be rebased now

**Review by georgebabey — APPROVED** [review]

**Review by georgebabey — APPROVED** [review]

**Pull request merged**